pkg/controller/namespace_helper.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2023 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+
+	"github.com/tigera/operator/pkg/common"
+	"github.com/tigera/operator/pkg/controller/utils"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func NewSingleTenantNamespaceHelper(ns string) NamespaceHelper {
+	return &namespacer{
+		multiTenant:           false,
+		singleTenantNamespace: ns,
+	}
+}
+
+func NewNamespaceHelper(mt bool, singleTenantNS, multiTenantNS string) NamespaceHelper {
+	return &namespacer{
+		multiTenant:           mt,
+		singleTenantNamespace: singleTenantNS,
+		multiTenantNamespace:  multiTenantNS,
+	}
+}
+
+type NamespaceHelper interface {
+	// InstallNamespace returns the namespace that components will be installed into.
+	// for single-tenant clusters, this is tigera-manager. For multi-tenancy, this
+	// will be the tenant's namespace.
+	InstallNamespace() string
+
+	// TruthNamespace returns the namespace to use as the source of truth for storing data.
```

Comment:
I've always thought of this namespace as the place to read user provided data from. Mostly secrets, as other input is typically and ideally in our CRDs.
I still don't like the name  TruthNamespace, I cannot (yet) get used to it.

Reply:
I think it's the best description of the purpose of the tigera-operator namespace. It's the namespace users can put configuration, sure. But it's also the namespace that we populate if the user didn't provide any information, and that information always flows from tigera-operator -> any namespace that needs it.

Controllers compare the value of secrets in the tigera-operator namespace with the "live" version of that secret in the namespace, and if they don't match they replace it with the one from the tigera-operator namespace. It really is the "source of truth" namespace.